lib/Core/SQLiteBuildDB.cpp

```diff
@@ -598,15 +602,26 @@ class SQLiteBuildDB : public BuildDB {
     return true;
   }
 
-  virtual void buildStarted() override {
+  virtual bool buildStarted(std::string *error_out) override {
     // Execute the entire build inside a single transaction.
     //
     // FIXME: We should revist this, as we probably wouldn't want a crash in the
     // build system to totally lose all build results.
-    int result = sqlite3_exec(db, "BEGIN IMMEDIATE;", nullptr, nullptr,
-        nullptr);
-    assert(result == SQLITE_OK);
-    (void)result;
+    char *cError;
+    int result = sqlite3_exec(db, "PRAGMA locking_mode = EXCLUSIVE", nullptr, nullptr, &cError);
+
+    if (result == SQLITE_OK) {
+      result = sqlite3_exec(db, "BEGIN EXCLUSIVE;", nullptr, nullptr, &cError);
+    }
+
+    if (result != SQLITE_OK) {
+      *error_out = (std::string("unable to initialize database (") + cError + ")");
+      sqlite3_free(cError);
+      sqlite3_close(db);
```

Comment:
I don't think this is safe, we shouldn't free the database here because the destructor is also going to free it? I think it is OK to use return an error -- the client can always retry later and things might work?